homeassistant/components/proximity/coordinator.py

```diff
@@ -69,189 +83,224 @@ def __init__(
         ):
-            _LOGGER.debug("no lat and lon in old_state -> unknown")
-            return {
-                "dist_to_zone": round(distances_to_zone[state_change_data.entity_id]),
-                "dir_of_travel": "unknown",
-                "nearest": entity_name,
-            }
-
-        # Reset the variables
-        distance_travelled: float = 0
+            return None
 
-        # Calculate the distance travelled.
         old_distance = distance(
-            proximity_latitude,
-            proximity_longitude,
-            state_change_data.old_state.attributes[ATTR_LATITUDE],
-            state_change_data.old_state.attributes[ATTR_LONGITUDE],
+            zone.attributes[ATTR_LATITUDE],
+            zone.attributes[ATTR_LONGITUDE],
+            old_latitude,
+            old_longitude,
         )
         new_distance = distance(
-            proximity_latitude,
-            proximity_longitude,
-            state_change_data.new_state.attributes[ATTR_LATITUDE],
-            state_change_data.new_state.attributes[ATTR_LONGITUDE],
+            zone.attributes[ATTR_LATITUDE],
+            zone.attributes[ATTR_LONGITUDE],
+            new_latitude,
+            new_longitude,
         )
-        assert new_distance is not None and old_distance is not None
+
+        # it is ensured, that distance can't be None, since zones must have lat/lon coordinates
+        assert old_distance is not None
+        assert new_distance is not None
         distance_travelled = round(new_distance - old_distance, 1)
 
-        # Check for tolerance
         if distance_travelled < self.tolerance * -1:
-            direction_of_travel = "towards"
-        elif distance_travelled > self.tolerance:
-            direction_of_travel = "away_from"
-        else:
-            direction_of_travel = "stationary"
-
-        # Update the proximity entity
-        dist_to: float | str
-        if dist_to_zone is not None:
-            dist_to = round(dist_to_zone)
-        else:
-            dist_to = DEFAULT_DIST_TO_ZONE
-
-        _LOGGER.debug(
-            "%s updated: distance=%s: direction=%s: device=%s",
-            self.friendly_name,
-            dist_to,
-            direction_of_travel,
-            entity_name,
-        )
+            return "towards"
+
+        if distance_travelled > self.tolerance:
+            return "away_from"
+
+        return "stationary"
+
+    async def _async_update_data(self) -> ProximityData:
+        """Calculate Proximity data."""
+        if (zone_state := self.hass.states.get(f"zone.{self.proximity_zone}")) is None:
+            _LOGGER.debug(
+                "%s: zone %s not exists -> reset",
+                self.friendly_name,
+                self.proximity_zone,
+            )
+            return DEFAULT_DATA
+
+        entities_data: dict[str, dict[str, str | int | None]] = self.data.entities
+
+        # calculate distance for all tracked entities
+        for device in self.proximity_devices:
```

Comment:
Maybe we should be consistent how we call the tracked entities? I'd prefer to call them just that, `tracked_entities`.

Reply:
8f88867